### DISCUSSION TITLE
Commercial Ecosystem

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#1598] [General]
In #732 I described that one of the most important aspects in the long-term success of DNN has been its commercial ecosystem. Essentially a commercial ecosystem is made possible by three fundamental capabilities: a platform, an extensibility model for creating and integrating custom features, and a distribution service for making the custom features easily discoverable and consumable by end users and monetizable by developers. Up until recently Oqtane mostly focused on the platform and extensibility model but as these have matured it has become time to focus on the distribution service.

Prior to the 2.2 release, Oqtane had only a basic integration with Nuget.org. Nuget.org was never designed to be a catalog service for a third party application such as Oqtane, so the capability was very limited. A more robust solution required the creation of a dedicated catalog service for Oqtane which was discussed in the #1277 RFC. 

In the 2.2 release a catalog service was created at www.oqtane.net and was integrated with the Oqtane framework. The catalog service will provide a convenient way for developers to register their products and for consumers to discover, download and install them into the framework.

The catalog service provides the underlying foundation to enable a commercial ecosystem. However, more services are needed and the longer term goal is to embrace the Oqtane philosophy of "speed" and provide developers with all of the essential capabilities for creating and managing commercial products. 

One of the core requirements for any commercial product is a technical licensing solution. Obviously there are many solutions available however integration is often a complicated and costly process. In addition, if each developer adopts their own licensing solution, it results in an inconsistent user experience for end users. To address these issues, Oqtane endeavors to provide a flexible licensing solution that developers can utilize to provide a turn-key experience. A prototype solution has been created and the expected functionality is described below...

Since Oqtane is based on Blazor it makes sense to provide a licensing component which developers can utilize within their products. A LicenseView component has been developed which is modeled after the Blazor AuthorizeView component. It has 3 different UI "states" - Licensing, Licensed, and NotLicensed - and it relies on a LicensingService for license validation which always occurs on the server. There is one required parameter of PackageName which refers to the unique Nuget PackageId for the product. Note that this LicenseView component is intended to be used for either Oqtane modules or themes. 

![image](https://user-images.githubusercontent.com/4840590/128715363-5969f9a0-8bfe-4143-9b29-0554356fa24d.png)

The default behavior of the LicenseView when a user does not have a valid license is that it will display a warning alert ( alongwith any other content the developer may have included in the NotLicensed section )

![image](https://user-images.githubusercontent.com/4840590/128715407-1f85d9f2-1c29-4468-8537-6f2f18e878bd.png)

If the user is a host, the alert message will contain some additional options for interacting with the licensing service including the ability to Purchase a license or Activate the product.

![image](https://user-images.githubusercontent.com/4840590/128715446-ecaf6e95-248b-452a-afc0-f8c6707c0a85.png)

Choosing the Activate option will allow the user to automatically retrieve a license key from a remote license server or enter a license key manually and save it.

![image](https://user-images.githubusercontent.com/4840590/128715483-52f377b8-3f86-4286-a44a-5d78553ab2d9.png)

The licensing service is based on an interface. The interface provides the flexibility for developers to provide their own license server for their products. A default license server is provided which is intended to meet the requirements of most developers and simplify the integration effort. The default license server relies on an API at www.oqtane.net for the initial license validation and license key creation, however for performance reasons it stores the license key locally on the web server ( as a *.lic file in the /bin folder ) for subsequent license validation operations.

![image](https://user-images.githubusercontent.com/4840590/128715520-3701d1b1-9c11-442d-ab89-b01f13c10251.png)

The catalog service on www.oqtane.net will provide the ability to specify the license information for products. It will include support for a free trial period, multiple license options and prices, multiple license periods ( perpetual, yearly, monthly ) and multiple license scopes ( unlimited, per installation, per domain ).

![image](https://user-images.githubusercontent.com/4840590/128715565-659bd801-2499-4a43-a71b-30ac2688fd20.png)

![image](https://user-images.githubusercontent.com/4840590/128715594-7899d1a4-b3c4-402e-8972-7d298ba47d07.png)

To complement the licensing service, there will obviously need to be an e-commerce capability as well so that users can purchase products. This is the purpose of the Purchase button in the licensing component UI included above. The concept is to use a familiar "app store" approach for purchasing products ( however the actual e-commerce transaction will need to take place on a secure remote server ). The default purchase process will be enabled on www.oqtane.net and will utilize the product information contained in the catalog service so that it only needs to be maintained in one location.

![image](https://user-images.githubusercontent.com/4840590/128715622-9f9d2a16-a00e-4335-8e1f-569bd315e3c0.png)

I would like to kick off a Beta phase for this capability soon but I am interested in hearing any feedback to the concept.

**Comment by mcownie** [comment]
i like the general concept alot, but in this day and age of data leaks and hacking would not an integration with a third party payment service(s) be 'safer' 
Personally I am pretty leary about putting my payment details into a website these days unless it whizzes me off to something like the ubiquitous paypal or recognizably kosher

**Reply by sbwalker** [reply]
@mcownie I 100% agree and plan to redirect to PayPal or Stripe for the actual payment processing.

**Comment by ADefWebserver** [comment]
My $0.02...

- www.oqtane.org can be the "Default" but, we must always allow other "E-Commerce Stores/Platforms" to be **_easily_** configured.

- If I have to move my Oqtane site, or restore it, I need to be able to get all my paid modules up and running again without the need to be connected to the internet (using .lic files downloaded to the "bin" directory would address this).

**Reply by sbwalker** [reply]
@ADefWebserver I agree with your points... the challenging thing is that the more flexibility you add to a system the harder it is to ensure a consistent user experience. For example, the goal is to make licensing flexible so that developers can plug-in in their own implementation - however it is possible that one of those implementations may rely on an Internet connection for license activation - which would mean that Oqtane could not guarantee that it could satisfy your second point above.

**Reply by ADefWebserver** [reply]
> one of those implementations may rely on an Internet connection for license activation - which would mean that Oqtane could not guarantee that it could satisfy your second point above.

Yeah I can see that. Perhaps just outlining a scenario (basically as 'guidance' for Module developers) would suffice. I would follow such guidance for my modules. Others (for example high priced modules) may decide differently.

**Reply by ADefWebserver** [reply]
> the more flexibility you add to a system the harder it is to ensure a consistent user experience.

An "alternative" license server would have to follow "**_exactly_**" the API of the default www.oqtane.org server. I was just thinking that on the Oqtane Install Settings page you could change that address from www.oqtane.org to something else if you wanted.

**Reply by sbwalker** [reply]
The problem is that APIs are just contracts - they cannot prescribe the underlying implementation. As long as they exactly meet the interface requirements, developers are free to do anything they want internally within their own implementations.  And just in case my earlier post was not clear, module licensing is completely independent from the Oqtane framework - there is no dependency. So it would not make sense to have any setting or UI option in the Oqtane framework itself.

**Reply by ADefWebserver** [reply]
> And just in case my earlier post was not clear, module licensing is completely independent from the Oqtane framework - there is no dependency. So it would not make sense to have any setting or UI option in the Oqtane framework itself.

I was thinking that clicking the "**Purchase**" button would simply pass the parameters to www.oqtane.org and it would display a purchase form? If so, it was the "www.oqtane.org" setting for "Store Address" that I was thinking about. Or am I still confused?

**Reply by sbwalker** [reply]
Perhaps what you are referring to is an alternate catalog service ( not license server )? Currently, the main goal of the catalog service is to allow users of Oqtane to easily discover and install extensions. Its possible that there may be a scenario where someone might want to create a private catalog service for their own private Oqtane installation which contains only a few curated extensions ( ie. a private Nuget server ). This scenario is currently not in scope - however someone could certainly create a custom module if they needed this functionality.

**Reply by ADefWebserver** [reply]
> Perhaps what you are referring to is an alternate catalog service ( not license server )?

No not an alternate Catalog server (someone can just post Oqtane modules on a website and people can download them and upload them into their Oqtane websites manually). I was thinking about where the "**Purchase**" button directs a user to.

**Reply by sbwalker** [reply]
The licensing component is independent from the framework. If you are building a commercial module, you could choose to utilize the licensing component as it provides a UI, services, etc... and an extensibility model for licensing providers. The payment url is provided by the licensing provider. The default licensing provider generates urls which point to oqtane.net. Developers can create their own licensing providers which generate other payment urls. However its important to note that there is much more involved than just providing a payment url - you also need to keep track of the order, expose an API to validate the order, generate a license key, etc...

**Reply by sbwalker** [reply]
Based on your scenario of making modules available for download from your own site, and having users upload them to their Oqtane install it sounds like you are not interested in the App Store concept - and that is totally fine... you do not need to register your modules in the catalog and you should still be able to use the licensing component ( but you would need to create a custom licensing provider ).

**Reply by ADefWebserver** [reply]
> Based on your scenario of making modules available for download from your own site, and having users upload them to their Oqtane install it sounds like you are not interested in the App Store concept - and that is totally fine... you do not need to register your modules in the catalog and you should still be able to use the licensing component ( but you would need to create a custom licensing provider ).

No I would rather use the Default option.

My only suggestion of making it "easily configurable" was only so that "Oqtane the company" could be 'aggressive' without the usual criticism that DNN corp got.

**Reply by sbwalker** [reply]
@ADefWebserver My main goal at this point is to construct a user experience which makes it as quick and seamless as possible for developers to sell extensions and users to consume them. I really appreciate your comments in this thread as I agree that there needs to be some alternate flows supported as well. So essentially there will be a "fully integrated" option that includes catalog, purchase, verification, delivery, and licensing aspects to provide a seamless end-to-end experience. And there will also be the ability to opt out of various aspects of this solution to provide flexibility.

**Comment by horacioj** [comment]
A (very welcome) commercial ecosystem will result in [this ](https://github.com/oqtane/oqtane.framework/discussions/1375) kind of problem being very common. 

An Oqtane installation could potentially be running many modules provided by different vendors. 
Each one of these modules could be referencing different versions of common (or not that common, whatever) NuGet packages and/or DLLs. 

This will lead undoubtedly to the typical "DLL hell". 

How could this be avoided, or at least be made less hellish avoiding having fragile Oqtane installations that could crash when installing or updating a module referencing an incompatible version of a library already installed by another module?

**Reply by sbwalker** [reply]
@horacioj you are correct that this is a potential side effect of having an ecosystem with many available extensions. In fact, this is an issue in DNN where there are thousands of available extensions - however the problem occurs much less frequently than one would expect. The solution is  to create a dependency manager... however this is a lower priority at this point in time.

**Comment by iJungleboy** [comment]
Just a minor issue: https://oqtane.net/ fails - should redirect to https://www.oqtane.net/

**Comment by iJungleboy** [comment]
@sbwalker Could you help? 

1. We're confused about how to get packages into this new installation system.   
  So ATM we have the image-resizer as well as 2sxc meant for Oqtane. Previously they appeared as expected, now using oqtane.net they don't appear any more. 
2. We also have the need to be able to specify compatibility versions. For example Oqtane 2.2 needs 2sxc 12.04 (just about to be released), other versions don't work. And Oqtane 2.1 can't run 2sxc 12.04, because it had to change for Oqtane 2.2. Will we be able to configure this somehow - in nuget or Oqtane.net?

Thx!

**Reply by sbwalker** [reply]
The registry service is still in beta so it has not been made publicly available on https://www.oqtane.net yet. It does support the ability to specify the compatibility of a product - so that it is able to filter the results to only those products which match the framework version installed. This is one of many reasons why Nuget.org was not an acceptable solution and a new registry service was needed. I can create an account for you so that you can register the 2SXC products.

**Reply by iJungleboy** [reply]
@sbwalker that would be awesome. We're working really hard to keep 2sxc in sync with Oqtane. You have my coordinates ;)

**Comment by sbwalker** [comment]
I would like to thank everyone for their feedback in this thread related to creating a technical foundation for a commercial ecosystem. The initial POC has evolved significantly and I would like to share the latest developments.

If you are following the Dev branch you will notice that there is some additional integration between the framework and the catalog service:

![image](https://user-images.githubusercontent.com/4840590/131187152-9795223d-6689-4ea9-9bce-371f722ad0df.png)

Highlighted in red are some new options related to commercial extensions. There is a download button which will only be displayed if the commercial extension offers a free trial period or the extension has been previously purchased. Clicking download now requires a user to accept the license terms before download:

![image](https://user-images.githubusercontent.com/4840590/131188271-f9eabfdc-4bbc-4fbc-a140-c13523195098.png)
 
There is also a button which displays a price and allows a user to navigate directly to a location where they can purchase the extension. The location is customizable and can be specified by the owner of the extension within the catalog registry. An integrated payment service will be available to simplify the process for those who do not want to set up their own service.

![image](https://user-images.githubusercontent.com/4840590/131187897-b4b0f682-a963-439b-82aa-437d166290cc.png)

The integrated payment service will allow the user to select their desired license and the actual payment processing will be handled by Stripe:

![image](https://user-images.githubusercontent.com/4840590/131187988-d8dd6156-a1e3-464f-8030-ac8d328ba589.png)

Once the purchase process is complete the user will be returned to their site where they can download the extension. Once the extension is downloaded and installed it will need to obtain a license. A licensing component will be available to developers who do not want to create their own licensing service - however it is not a requirement.

![image](https://user-images.githubusercontent.com/4840590/131188367-81fbddde-9998-4660-84dd-bd9c239c4672.png)

Clicking the Activate button will allow the user to retrieve a license key ( or enter it manually ):

![image](https://user-images.githubusercontent.com/4840590/131188485-eed1f462-336d-4939-a334-085caa536569.png)

Once the license key is validated the extension will be become functional. The default license server saves the license key locally on the server so that it does not have to phone home in order to do license validation. 

![image](https://user-images.githubusercontent.com/4840590/131188533-26c53dbb-9803-491f-899c-b8212120a06d.png)

The licensing component is very simple for a developer to use. They simply need to specify the PackageName property. They also have the ability to specify a PaymentUrl if they are using their own payment service, and they can specify their own LicenseServerType if they want to provide their own licensing server implementation. 

![image](https://user-images.githubusercontent.com/4840590/131188719-dcaa4fd1-56d4-4756-87a2-e1d221608378.png)

The LicenseServer interface has been changed from the earlier version to make it more intuitive and adaptable to future scenarios:

![image](https://user-images.githubusercontent.com/4840590/131188931-4f2f870e-ec25-4598-a465-58a26297da10.png)

The catalog registry UI has also been enhanced to make it more intuitive and provide the option to specify a custom payment service:

![image](https://user-images.githubusercontent.com/4840590/131189008-0eebc6e0-91f4-4317-a55a-88412cd01a99.png)

 And release management has been enhanced to provide an integrated download capability:

![image](https://user-images.githubusercontent.com/4840590/131189148-f3997f3a-492f-4ff4-aadd-6352dbdb8056.png)

Choosing to use the integrated payment, download, and licensing services will provide the most seamless user experience to users. However developers will be able to plug in their own services as well ( with the caveat that the user experience will not be as streamlined ).

I am interested in hearing any feedback or questions.

**Comment by mikesmeltzer** [comment]
This is great and will help reduce barriers to entry from a commercial perspective for developers @sbwalker. One question on the integrated payment service, how does the payment end up with the developer? Would my Stripe account be linked through the catalog registry or is payment sent to the developer by another means?

**Reply by sbwalker** [reply]
For the integrated payment service the payouts will be done via Stripe Connect ( https://stripe.com/connect ). It will also be possible for developers to specify their own payment service url... but if they choose this approach they will need to provide a solution for downloads and licensing, as the framework would not be able to verify the status of orders.

**Reply by sbwalker** [reply]
The other obvious question is related to commissions for the integrated payment service. The recent trend in the industry is to try to encourage more developers to create apps for platforms... and one of the strategies that companies such as Apple, Google, Amazon, and Shopify have used to accomplish this goal is lowering their commissions. Obviously these platforms have a massive user base already so they can rely on sales volume to offset the loss in commissions. However I think the strategy is sound regardless. So my preference is to minimize commissions to just cover costs so that the Oqtane ecosystem has the greatest potential for growth.

**Reply by horacioj** [reply]
What if Stripe isn't available in the developer's country? The only available path would be 'specify their own payment service url., etc.' or there could be a workaround?

**Reply by sbwalker** [reply]
@horacioj yes, in that situation the solution would be for the developer to provide their own payment service. Keep in mind that the integrated service is focused on an MVP which should accommodate the vast majority of developers and users - however it will certainly not handle every scenario - which is why we need to provide the flexibility for other options.

**Reply by horacioj** [reply]
😔 I've been waiting for years for Stripe to be available in my Country, and I think it will never happen. Uruguay is too small, and it is not worth their effort.

**Reply by sbwalker** [reply]
@horacioj the Stripe site indicates that they support payouts for Uruguay https://stripe.com/docs/connect/cross-border-payouts

**Reply by horacioj** [reply]
Cool! I don't need a Stripe account myself. Thank you, @sbwalker.

**Comment by mikesmeltzer** [comment]
@sbwalker If a link to the vendors / products website could be added to the UI within each module listing in the Module Management screen that would be great too (if not already there).

**Reply by sbwalker** [reply]
It already exists in Oqtane 2.2

**Comment by leigh-pointer** [comment]
@sbwalker  are you planning to update this to .net6 or are the .net5 still valid?

**Reply by sbwalker** [reply]
Yes, I plan to upgrade this to .NET 6 - and I also plan to expose the commercial aspects of this service very soon so that developers can sell extensions.
